File: libs/go/bullet.go
package stonecharts

import (
	"fmt"
	"math"
	"sort"
	"strings"
)

const bulletPad = 0.2
const bulletMeasureRatio = 0.4
const bulletTargetRatio = 0.6
const bulletTargetWidth = 2

var lightRangeShades = []string{"#cccccc", "#dddddd", "#eeeeee"}
var darkRangeShades = []string{"#3d3d55", "#2d2d42", "#1e1e30"}

func rangeFills(n int, isDark bool) []string {
	shades := lightRangeShades
	if isDark {
		shades = darkRangeShades
	}
	fills := make([]string, n)
	for k := 0; k < n; k++ {
		idx := k
		if idx >= len(shades) {
			idx = len(shades) - 1
		}
		fills[k] = shades[idx]
	}
	return fills
}

func renderBulletSVG(spec *ChartSpec) string {
	s2 := *spec
	s2.Series = make([]Series, len(spec.Series))
	copy(s2.Series, spec.Series)

	ranges := s2.BulletRanges
	target := s2.BulletTarget

	allVals := []float64{0.0}
	for _, s := range s2.Series {
		allVals = append(allVals, s.Data...)
	}
	if target != nil {
		allVals = append(allVals, *target)
	}
	allVals = append(allVals, ranges...)

	if s2.YAxis.Min == nil {
		mn := allVals[0]
		for _, v := range allVals[1:] {
			if v < mn {
				mn = v
			}
		}
		s2.YAxis.Min = &mn
	}
	if s2.YAxis.Max == nil {
		mx := allVals[0]
		for _, v := range allVals[1:] {
			if v > mx {
				mx = v
			}
		}
		s2.YAxis.Max = &mx
	}

	return renderCartesian(&s2, "Bullet", "band", bulletMarks, true, "horizontal")
}

func bulletMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}
	spec := f.spec
	ranges := spec.BulletRanges
	target := spec.BulletTarget
	isDark := f.theme.Name == "dark"
	targetColor := "#333333"
	if isDark {
		targetColor = "#cccccc"
	}

	bandHeight := f.bandHeight()
	groupH := bandHeight * (1 - bulletPad)
	stacked := f.stacking == "normal" || f.stacking == "percent"
	kSlots := len(f.spec.Series)
	if stacked || !f.spec.groupingOn() {
		kSlots = 1
	}
	if kSlots <= 0 {
		kSlots = 1
	}
	barH := groupH / float64(kSlots)
	measureH := barH * bulletMeasureRatio
	targetH := barH * bulletTargetRatio
	baseline := f.valueZero()

	if len(ranges) > 0 {
		sortedRanges := make([]float64, len(ranges))
		copy(sortedRanges, ranges)
		sort.Float64s(sortedRanges)
		fills := rangeFills(len(sortedRanges), isDark)
		for i := 0; i < f.n; i++ {
			cyBand := f.bandCenter(i)
			bandTop := cyBand - groupH/2
			prevX := baseline
			for k, rVal := range sortedRanges {
				rx := f.valuePix(rVal)
				x := math.Min(prevX, rx)
				w := math.Abs(rx - prevX)
				p.WriteString(fmt.Sprintf(
					`<rect class="sc-range" x="%s" y="%s" width="%s" height="%s" fill="%s"/>`,
					f1(x), f1(bandTop), f1(w), f1(groupH), fills[k]))
				prevX = rx
			}
		}
	}

	for si, s := range spec.Series {
		st := f.styles[si]
		p.WriteString(fmt.Sprintf(`<g class="sc-series" data-series="%d">`, si))
		for i, raw := range s.Data {
			if i >= f.n {
				break
			}
			cyBand := f.bandCenter(i)
			slot := 0
			if f.spec.groupingOn() && !stacked {
				slot = si
			}
			slotTop := cyBand - groupH/2 + barH*float64(slot)
			measureTop := slotTop + (barH-measureH)/2

			xv := f.valuePix(raw)
			x := math.Min(baseline, xv)
			w := math.Abs(baseline - xv)
			tip := xv
			cy := measureTop + measureH/2

			ylabel := fmt.Sprintf("%d", i)
			if i < len(f.cats) {
				ylabel = f.cats[i]
			}
			common := fmt.Sprintf(
				`class="sc-bar sc-point" data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-r="3.5" data-r-hover="6"`,
				si, esc(s.Name), esc(ylabel), esc(fmtNum(raw)), st.solid)
			p.WriteString(fmt.Sprintf(
				`<rect %s cx="%s" cy="%s" x="%s" y="%s" width="%s" height="%s" fill="%s"/>`,
				common, f1(tip), f1(cy), f1(x), f1(measureTop), f1(w), f1(measureH), st.fill))
		}
		p.WriteString(`</g>`)
	}

	if target != nil {
		tv := *target
		for i := 0; i < f.n; i++ {
			cyBand := f.bandCenter(i)
			targetTop := cyBand - targetH/2
			tx := f.valuePix(tv)
			p.WriteString(fmt.Sprintf(
				`<line class="sc-target" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s" stroke-width="%d"/>`,
				f1(tx), f1(targetTop), f1(tx), f1(targetTop+targetH), targetColor, bulletTargetWidth))
		}
	}
}

File: libs/python/stonecharts/charts/bullet.py
"""Bullet chart renderer: ChartSpec -> SVG string.

Horizontal KPI bars with a comparative target tick and qualitative range bands.
Shared Cartesian chrome comes from _cartesian.py; this module draws only the
bullet marks. See charts/bullet/design.md for the full geometry contract.
"""

from __future__ import annotations

import copy

from ..spec import ChartSpec
from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, render_cartesian

PAD = 0.2
MEASURE_RATIO = 0.4
TARGET_RATIO = 0.6
TARGET_WIDTH = 2

_RANGE_SHADES_LIGHT = ["#cccccc", "#dddddd", "#eeeeee"]
_RANGE_SHADES_DARK = ["#3d3d55", "#2d2d42", "#1e1e30"]


def _range_fills(n: int, is_dark: bool) -> list[str]:
    shades = _RANGE_SHADES_DARK if is_dark else _RANGE_SHADES_LIGHT
    if n <= len(shades):
        return shades[:n]
    return [shades[min(k, len(shades) - 1)] for k in range(n)]


def render_svg(spec: ChartSpec) -> str:
    mod = copy.copy(spec)
    mod.y_axis = copy.copy(spec.y_axis)

    ranges = getattr(mod, "bullet_ranges", None) or []
    target = getattr(mod, "bullet_target", None)

    all_vals = [0.0]
    for s in mod.series:
        all_vals.extend(s.data)
    if target is not None:
        all_vals.append(target)
    all_vals.extend(ranges)

    if mod.y_axis.min is None:
        mod.y_axis.min = min(all_vals)
    if mod.y_axis.max is None:
        mod.y_axis.max = max(all_vals)

    return render_cartesian(mod, "Bullet", "band", _bullet_marks, orientation="horizontal")


def _bullet_marks(fr: CartesianFrame, p: list) -> None:
    if fr.n <= 0:
        return

    spec = fr.spec
    ranges = getattr(spec, "bullet_ranges", None) or []
    target = getattr(spec, "bullet_target", None)
    is_dark = fr.theme.name == "dark"
    target_color = "#cccccc" if is_dark else "#333333"

    band_height = fr.band_height()
    group_h = band_height * (1 - PAD)
    stacked = fr.stacking in ("normal", "percent")
    k_slots = 1 if stacked or not fr.spec.grouping else max(len(fr.spec.series), 1)
    bar_h = group_h / k_slots
    measure_h = bar_h * MEASURE_RATIO
    target_h = bar_h * TARGET_RATIO
    baseline = fr.value_zero()

    if ranges:
        sorted_ranges = sorted(ranges)
        fills = _range_fills(len(sorted_ranges), is_dark)
        for i in range(fr.n):
            cy_band = fr.band_center(i)
            band_top = cy_band - group_h / 2
            prev_x = baseline
            for k, r_val in enumerate(sorted_ranges):
                rx = fr.value_pix(r_val)
                x = min(prev_x, rx)
                w = abs(rx - prev_x)
                p.append(
                    f'<rect class="sc-range" x="{x:.1f}" y="{band_top:.1f}" '
                    f'width="{w:.1f}" height="{group_h:.1f}" fill="{fills[k]}"/>'
                )
                prev_x = rx

    for si, s in enumerate(spec.series):
        st = fr.styles[si]
        p.append(f'<g class="sc-series" data-series="{si}">')
        for i, raw in enumerate(s.data):
            if i >= fr.n:
                break
            cy_band = fr.band_center(i)
            slot = si if fr.spec.grouping and not stacked else 0
            slot_top = cy_band - group_h / 2 + bar_h * slot
            measure_top = slot_top + (bar_h - measure_h) / 2

            xv = fr.value_pix(raw)
            x = min(baseline, xv)
            w = abs(baseline - xv)
            tip = xv
            cy = measure_top + measure_h / 2

            ylabel = fr.cats[i] if i < len(fr.cats) else str(i)
            common = (
                f'class="sc-bar sc-point" data-series="{si}" data-series-name="{esc(s.name)}" '
                f'data-x="{esc(ylabel)}" data-y="{esc(fmt_num(raw))}" '
                f'data-color="{st.solid}" data-r="3.5" data-r-hover="6"'
            )
            p.append(
                f'<rect {common} cx="{tip:.1f}" cy="{cy:.1f}" x="{x:.1f}" y="{measure_top:.1f}" '
                f'width="{w:.1f}" height="{measure_h:.1f}" fill="{st.fill}"/>'
            )
        p.append("</g>")

    if target is not None:
        for i in range(fr.n):
            cy_band = fr.band_center(i)
            target_top = cy_band - target_h / 2
            tx = fr.value_pix(target)
            p.append(
                f'<line class="sc-target" x1="{tx:.1f}" y1="{target_top:.1f}" '
                f'x2="{tx:.1f}" y2="{target_top + target_h:.1f}" '
                f'stroke="{target_color}" stroke-width="{TARGET_WIDTH}"/>'
            )
